### PR TITLE
2026 - Updated to support spying all methods on an object

### PR DESCRIPTION
**Pull request by benawhite**
#### Purpose
Fix issue #2026 by adding **spyEntireObject** to the supported signature of sinon.spy
added tests and updated docs to reflect the change

 #### How to verify
1. Check out this branch
2. `npm install`
3. npm run test
4. note: **assert-test** / **supports no properties** uses the newly added syntax and no other tests are broken
```
sinonSpy(api);
```

 #### Checklist for author

- [X] `npm run lint` passes
- [X] References to standard library functions are [cached](https://github.com/sinonjs/sinon/pull/1523).

### CODE REVIEW

File: lib/sinon/spy-entire-object.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+"use strict";
+
+var getPropertyDescriptor = require("./util/core/get-property-descriptor");
+var walk = require("./util/core/walk");
+
+function spyEntireObject(spy, object) {
+    walk(object || {}, function(prop, propOwner) {
+        // we don't want to spy things like toString(), valueOf(), etc. so we only spy if the object
```
Comment (lucasfcosta) [inline lib/sinon/spy-entire-object.js]: Is there any particular reason for us no to spy on those? IMO these should also be spied on; otherwise it feels like we're making decisions on behalf of our users. I agree most users wouldn't need these spied on, but that's up to them.

I'd go for the least astonishment principle here and spy on every single function.

People can also be attaching methods to `Object.prototype` — which is weird, but valid nonetheless.

File: lib/sinon/spy-entire-object.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+"use strict";
+
+var getPropertyDescriptor = require("./util/core/get-property-descriptor");
+var walk = require("./util/core/walk");
+
+function spyEntireObject(spy, object) {
+    walk(object || {}, function(prop, propOwner) {
```
Comment (lucasfcosta) [inline lib/sinon/spy-entire-object.js]: Maybe we could have this callback as an arrow function?

File: lib/sinon/spy-entire-object.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        // we don't want to spy things like toString(), valueOf(), etc. so we only spy if the object
+        // is not Object.prototype
+        if (
+            propOwner !== Object.prototype &&
+            prop !== "constructor" &&
```
Comment (lucasfcosta) [inline lib/sinon/spy-entire-object.js]: We may also want to spy on the `constructor` for the same reasons above. I think reducing edge-cases would be productive here.

File: docs/release-source/release/spies.md [hunk 4]
```diff
@@ -32,6 +32,40 @@ handles a callback, as in the following simplified example:
...
 
+### Using a spy to wrap all object method
+
+`sinon.spy(object)`
+
+Spies all the object's methods.
+
+Note that it's usually better practice to spy individual methods, particularly on objects that you don't understand or control all the methods for (e.g. library dependencies).
+
+spying individual methods tests intent more precisely and is less susceptible to unexpected behavior as the object's code evolves.
```
Comment (lucasfcosta) [inline docs/release-source/release/spies.md]: Sorry for nitpicking, but I think the first character here needs to be in uppercase.

File: docs/release-source/release/spies.md [hunk 5]
```diff
@@ -32,6 +32,40 @@ handles a callback, as in the following simplified example:
 }
 ```
 
+### Using a spy to wrap all object method
```
Comment (lucasfcosta) [inline docs/release-source/release/spies.md]: Very well written docs, I especially like the example, thanks!

File: lib/sinon/spy-entire-object.js (hunk 2 above)
Comment (fatso83) [reply lib/sinon/spy-entire-object.js]: the entire codebase is ES5.1, so arrow functions are out.

File: lib/sinon/spy-entire-object.js (hunk 3 above)
Comment (benawhite) [reply lib/sinon/spy-entire-object.js]: This is a blatant copy of stub-entire-object.js as I wanted to keep the API for the two in sync.

Comment (benawhite) [reply lib/sinon/spy-entire-object.js]: FYI, I also agree with the way it was originally written. Spying on native object methods seems out of scope...

Comment (lucasfcosta) [reply lib/sinon/spy-entire-object.js]: I see. That's fine then, I didn't know that, sorry 😅

**Comment by fatso83** [comment]
This is all very nice work, @benawhite. Any chance of getting those ✅ nailed down before the Summer gets to us all? 🌞

**Comment by fatso83** [comment]
Having a new look at this, seeing it has been stalling for two months. Right now, there are two things I don't understand:
1. Which tests/edge cases are you pointing to, @mantoni? I scanned the changed files several times without seeing any todos or special comments. The only thing I see that hasn't been adressed is a minor spelling/letter casing thing mentioned by Lucas.
2. What is the fourth point in the PR description really saying, @benawhite? I don't understand this:
> note: assert-test / supports no properties uses the newly added syntax and no other tests are broken

I could fix the spelling issue and merge this if someone could enlighten me on the above two points.

**Comment by mantoni** [comment]
There is a comment saying

> we don't want to spy things like toString(), valueOf(), etc.

It‘s documenting and if statement. There are no tests covering these cases.

Sorry, I should have been more specific.

**Comment by fearphage** [comment]
As it stands, you either have to use this with a `sandbox` or you have to go to each property of the related object to call `.restore()`. It would be nice (read: good user experience) if `spyOnEntireObj` returned a way to revert everything back to normal.

**Comment by fatso83** [comment]
I created a new branch that addresses all of the points above, except Phred's suggestion. I feel that improvement belongs in a separate PR, as it should touch both the stubs and spies API.

@mantoni, you'll find the tests you requested as `does not replace non-enumerable properties from Object.prototype`, which are now shared between the stubs and spies test suite.
